File: progress.py
"""События прогресса для live-лога синка/upload."""

from __future__ import annotations

from typing import Any, Callable, Literal, TypedDict


ProgressStage = Literal[
    "start",
    "auth",
    "list",
    "fetch",
    "classify",
    "parse",
    "build",
    "done",
    "error",
    "warn",
]


class ProgressEvent(TypedDict, total=False):
    stage: ProgressStage
    message: str
    current: int
    total: int
    company: str
    detail: str


ProgressCb = Callable[[ProgressEvent], None]


def emit(
    cb: ProgressCb | None,
    stage: ProgressStage,
    message: str,
    **extra: Any,
) -> None:
    if cb is None:
        return
    event: ProgressEvent = {"stage": stage, "message": message}
    for key, value in extra.items():
        if value is not None:
            event[key] = value  # type: ignore[literal-required]
    cb(event)

File: web/src/types/progress.ts
export type ProgressStage =
  'start' | 'auth' | 'list' | 'fetch' | 'classify' | 'parse' | 'build' | 'done' | 'error' | 'warn'

export type ProgressEvent = {
  type?: 'progress' | 'done' | 'error'
  stage?: ProgressStage
  message: string
  current?: number
  total?: number
  company?: string
  detail?: string
  sessionId?: string
  status?: number
  report?: import('./report').Report
}

export const PROGRESS_STAGES: Array<{
  id: ProgressStage
  label: string
}> = [
  { id: 'auth', label: 'Сессия' },
  { id: 'list', label: 'Список чатов' },
  { id: 'fetch', label: 'Переписки' },
  { id: 'classify', label: 'Классификация' },
  { id: 'parse', label: 'Разбор файла' },
  { id: 'build', label: 'Отчёт' },
]
